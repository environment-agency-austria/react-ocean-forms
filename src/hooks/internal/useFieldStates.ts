import { useRef, useCallback } from 'react';

import { IValidationArgs, IBasicValidationState, TUpdateMethod } from '../useValidation';

/**
 * Interface describing field states
 */
export interface IFieldState {
  /**
   * Label of the field
   */
  label: string;
  /**
   * True if the field is actually a FieldGroup
   */
  isGroup?: boolean;
  /**
   * Triggers the validation of this field
   * @param args Validation args
   */
  validate(args?: Partial<IValidationArgs>): Promise<IBasicValidationState>;
  /**
   * Returns the current value of the field
   */
  getValue(): unknown;
  /**
   * Resets the field to its initial state
   */
  reset(): void;
  /**
   * Updates the validation state of the field
   * @param state New validation state
   */
  updateValidation: TUpdateMethod;
}

export interface IUseFieldStatesResult {
  getFieldState(name: string): IFieldState;
  registerField(name: string, fieldState: IFieldState): void;
  unregisterField(name: string): void;
  forEachFieldState(callback: (value: IFieldState, key: string, map: Map<string, IFieldState>) => void): void;
}

export function useFieldStates(): IUseFieldStatesResult {
  const fields = useRef(new Map<string, IFieldState>());

  /**
   * Returns the current state of the given field
   * @param name Field name
   * @returns Current field state or default field state
   */
  const getFieldState = useCallback((name: string): IFieldState => {
    const fieldState = fields.current.get(name);
    if (fieldState === undefined) {
      throw new Error(`[Form] getFieldState: Could not find state of field '${name}'`);
    }

    return fieldState;
  }, []);

  /**
   * Registers a new field to the form.
   * @param name Field name
   * @param fieldState Field state
   */
  const registerField = useCallback((name: string, fieldState: IFieldState): void => {
    if (typeof name !== 'string' || name.length === 0) {
      throw new Error('[Form] registerField: name is required');
    }

    if (typeof fieldState !== 'object') {
      throw new Error('[Form] registerField: field state is required');
    }

    if (
      typeof fieldState.label !== 'string'
      || typeof fieldState.validate !== 'function'
      || typeof fieldState.updateValidation !== 'function'
      || typeof fieldState.reset !== 'function'
      || typeof fieldState.getValue !== 'function'
    ) {
      throw new Error('[Form] registerField: invalid field state given');
    }

    if (fields.current.has(name)) {
      throw new Error(`[Form] registerField: Could not register field with name '${name}'. A field with this name already exists inside this form.`);
    }

    fields.current.set(name, fieldState);
  }, []);

  /**
   * Unregisters a field from the form.
   * @param name Field name
   */
  const unregisterField = useCallback((name: string): void => {
    fields.current.delete(name);
  }, []);

  const forEachFieldState = useCallback((callback: (value: IFieldState, key: string, map: Map<string, IFieldState>) => void) => {
    return fields.current.forEach(callback);
  }, []);

  return {
    getFieldState,
    registerField,
    unregisterField,
    forEachFieldState,
  };
}
