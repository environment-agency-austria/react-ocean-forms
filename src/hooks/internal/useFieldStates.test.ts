import { renderHook } from '@testing-library/react-hooks';

import { useFieldStates, IFieldState, IUseFieldStatesResult } from './useFieldStates';

const createMockFieldState = (label: string, isGroup?: boolean): IFieldState => ({
  label,
  isGroup,
  validate: jest.fn().mockResolvedValue({
    isValidating: false,
    valid: true,
    error: null,
  }),
  updateValidation: jest.fn(),
  reset: jest.fn(),
  getValue: jest.fn().mockReturnValue(label),
});

interface IMockField {
  name: string;
  state: IFieldState;
}

const createMockField = (name: string, label: string, isGroup?: boolean): IMockField => ({
  name,
  state: createMockFieldState(label, isGroup),
});

describe('useFieldStates', () => {
  interface ISetupResult {
    result: {
      current: IUseFieldStatesResult;
    };
  }

  const setup = (): ISetupResult => renderHook(useFieldStates);

  it('should return the correct result', () => {
    const { result } = setup();

    expect(result.current).toMatchObject({
      getFieldState: expect.any(Function),
      registerField: expect.any(Function),
      unregisterField: expect.any(Function),
      forEachFieldState: expect.any(Function),
    });
  });

  describe('invalid field registration', () => {
    const mf = (): void => {};
    const cases: any[] = [
      ['no parameters', undefined, undefined],
      ['invalid field name', '', undefined],
      ['no state', 'foo', undefined],
      ['empty state', 'foo', {}],
      ['1 of 5 props', 'foo', { label: 'hey' }],
      ['2 of 5 props', 'foo', { label: 'hey', validate: mf }],
      ['3 of 5 props', 'foo', { label: 'hey', validate: mf, updateValidation: mf }],
      [
        '4 of 5 props',
        'foo',
        {
          label: 'hey',
          validate: mf,
          updateValidation: mf,
          reset: mf,
        },
      ],
    ];

    test.each(cases)('case %s', (testName: string, fieldName: string, fieldState: IFieldState) => {
      const { result } = setup();
      expect(() => {
        result.current.registerField(fieldName, fieldState);
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('field registration and values', () => {
    const createCases = (): [string, IMockField, boolean?][] => {
      return [
        ['field', createMockField('unitField', 'Unit field')],
        ['group', createMockField('unitGroup', 'Unit group', true)],
        ['sub field', createMockField('unitGroup.subField', 'Sub field', true)],
      ];
    };

    describe('registerField - field registration', () => {
      const cases = createCases();

      test.each(cases)(
        'should register a new %s without crashing',
        (testName, field: IMockField) => {
          const { result } = setup();
          expect(() => {
            result.current.registerField(field.name, field.state);
          }).not.toThrowError();
        }
      );

      it('should throw an error if a field with the given name is already registered', () => {
        const { result } = setup();

        const mockName = 'doubleField';
        const { state: mockState } = createMockField(mockName, mockName);

        result.current.registerField(mockName, mockState);

        expect(() => {
          result.current.registerField(mockName, mockState);
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('unregisterField - field cleanup', () => {
      const cases = createCases();

      test.each(cases)('should unregister %s without crashing', (testName, field: IMockField) => {
        const { result } = setup();
        result.current.registerField(field.name, field.state);
        result.current.unregisterField(field.name);
      });
    });

    describe('getFieldState - field states', () => {
      const cases = createCases();

      test.each(cases)(
        'should return the correct field state of a %s',
        (testName, field: IMockField) => {
          const { result } = setup();
          result.current.registerField(field.name, field.state);
          expect(result.current.getFieldState(field.name)).toBe(field.state);
        }
      );

      it('should throw an error when trying to access an non-existing field state', () => {
        const { result } = setup();
        const mockFieldName = 'mock-test';

        expect(() => result.current.getFieldState(mockFieldName)).toThrowError(
          `[Form] getFieldState: Could not find state of field '${mockFieldName}'`
        );
      });
    });

    it('forEachFieldState - should correctly iterate through the field states', () => {
      const { result } = setup();

      const field = createMockField('unitField', 'Unit field');
      result.current.registerField(field.name, field.state);

      const group = createMockField('unitGroup', 'Unit group', true);
      result.current.registerField(group.name, group.state);

      const subField = createMockField('unitGroup.subField', 'Sub field', true);
      result.current.registerField(subField.name, subField.state);

      const forEachCallback = jest.fn();
      result.current.forEachFieldState(forEachCallback);

      expect(forEachCallback).toHaveBeenCalledTimes(3);
      expect(forEachCallback).toHaveBeenNthCalledWith(
        1,
        field.state,
        field.name,
        expect.anything()
      );
      expect(forEachCallback).toHaveBeenNthCalledWith(
        2,
        group.state,
        group.name,
        expect.anything()
      );
      expect(forEachCallback).toHaveBeenNthCalledWith(
        3,
        subField.state,
        subField.name,
        expect.anything()
      );
    });
  });
});
