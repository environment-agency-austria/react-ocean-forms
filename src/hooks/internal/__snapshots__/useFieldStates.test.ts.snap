// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFieldStates field registration and values registerField - field registration should throw an error if a field with the given name is already registered 1`] = `"[Form] registerField: Could not register field with name 'doubleField'. A field with this name already exists inside this form."`;

exports[`useFieldStates invalid field registration case 1 of 5 props 1`] = `"[Form] registerField: invalid field state given"`;

exports[`useFieldStates invalid field registration case 2 of 5 props 1`] = `"[Form] registerField: invalid field state given"`;

exports[`useFieldStates invalid field registration case 3 of 5 props 1`] = `"[Form] registerField: invalid field state given"`;

exports[`useFieldStates invalid field registration case 4 of 5 props 1`] = `"[Form] registerField: invalid field state given"`;

exports[`useFieldStates invalid field registration case empty state 1`] = `"[Form] registerField: invalid field state given"`;

exports[`useFieldStates invalid field registration case invalid field name 1`] = `"[Form] registerField: name is required"`;

exports[`useFieldStates invalid field registration case no parameters 1`] = `"[Form] registerField: name is required"`;

exports[`useFieldStates invalid field registration case no state 1`] = `"[Form] registerField: field state is required"`;
